### PR TITLE
Refactor parse_and_execute

### PR DESCRIPTION
**Pull request by tomaka**
Deprecates `parse_and_execute` and replaces it with a new `parse_and_prepare` function whose control flow is in my opinion way easier to read.

I invite you to check out the third commit to see what the new code looks like.

The reason for this change is that I'd like to get rid of the `ServiceFactory` and `Components` traits, which `parse_and_execute` uses.
This PR instead moves the `ServiceFactory` trait usage to when the user calls `run()`, which makes it possible in the future to add functions other than `run()` that don't require this trait.
As an example, we could add a function to `ParseAndExecute::RevertChain` that simple requires passing a `Client` by parameter, instead of asking for a `ServiceFactory`.

### CODE REVIEW

**Comment by tomaka** [comment]
Another possible change in the future would be to add new functions to `ParseAndPrepareRun`, `ParseAndPrepareImport` and `ParseAndPrepareExport` that return a `Future`, thus letting the user control the blocks import or export and removing the `exit` parameter.

File: core/cli/src/lib.rs [hunk 1]
```diff
@@ -222,23 +213,294 @@ where
...
+		spec_factory: S,
+		exit: E,
+		run_service: RS,
+	) -> error::Result<()>
+	where S: FnOnce(&str) -> Result<Option<ChainSpec<G>>, String>,
+		RP: StructOpt + Clone,
+		C: Default,
+		G: RuntimeGenesis,
+		E: IntoExit,
+		RS: FnOnce(E, RunCmd, RP, Configuration<C, G>) -> Result<(), String> {
```
Comment (bkchr) [inline core/cli/src/lib.rs]:
```suggestion
		RS: FnOnce(E, RunCmd, RP, Configuration<C, G>) -> Result<(), String>
	{
```

File: core/cli/src/lib.rs [hunk 2]
```diff
@@ -222,23 +213,294 @@ where
...
+
```
Comment (bkchr) [inline core/cli/src/lib.rs]:
```suggestion
```

File: core/cli/src/lib.rs [hunk 3]
```diff
@@ -222,23 +213,294 @@ where
...
+}
+
+impl<'a> ParseAndPrepareBuildSpec<'a> {
+	/// Runs the command and build the chain specs.
+	pub fn run<G, S>(
+		self,
+		spec_factory: S
+	) -> error::Result<()>
+	where S: FnOnce(&str) -> Result<Option<ChainSpec<G>>, String>,
+		G: RuntimeGenesis {
```
Comment (bkchr) [inline core/cli/src/lib.rs]:
```suggestion
		G: RuntimeGenesis
	{
```

File: core/cli/src/lib.rs [hunk 4]
```diff
@@ -222,23 +213,294 @@ where
...
+
```
Comment (bkchr) [inline core/cli/src/lib.rs]:
```suggestion
```

File: core/cli/src/lib.rs [hunk 5]
```diff
@@ -222,23 +213,294 @@ where
...
+impl<'a> ParseAndPrepareExport<'a> {
+	/// Runs the command and exports from the chain.
+	pub fn run<F, S, E>(
+		self,
+		spec_factory: S,
+		exit: E,
+	) -> error::Result<()>
+	where S: FnOnce(&str) -> Result<Option<ChainSpec<FactoryGenesis<F>>>, String>,
+		F: ServiceFactory,
+		E: IntoExit {
```
Comment (bkchr) [inline core/cli/src/lib.rs]:
```suggestion
		E: IntoExit
	{
```

File: core/cli/src/lib.rs [hunk 6]
```diff
@@ -222,23 +213,294 @@ where
...
+
```
Comment (bkchr) [inline core/cli/src/lib.rs]:
```suggestion
```

File: core/cli/src/lib.rs [hunk 7]
```diff
@@ -222,23 +213,294 @@ where
...
+impl<'a> ParseAndPrepareImport<'a> {
+	/// Runs the command and imports to the chain.
+	pub fn run<F, S, E>(
+		self,
+		spec_factory: S,
+		exit: E,
+	) -> error::Result<()>
+	where S: FnOnce(&str) -> Result<Option<ChainSpec<FactoryGenesis<F>>>, String>,
+		F: ServiceFactory,
+		E: IntoExit {
```
Comment (bkchr) [inline core/cli/src/lib.rs]:
```suggestion
		E: IntoExit
	{
```

File: core/cli/src/lib.rs [hunk 8]
```diff
@@ -222,23 +213,294 @@ where
...
+
```
Comment (bkchr) [inline core/cli/src/lib.rs]:
```suggestion
```

File: core/cli/src/lib.rs [hunk 9]
```diff
@@ -222,23 +213,294 @@ where
...
+}
+
+impl<'a> ParseAndPreparePurge<'a> {
+	/// Runs the command and purges the chain.
+	pub fn run<G, S>(
+		self,
+		spec_factory: S
+	) -> error::Result<()>
+	where S: FnOnce(&str) -> Result<Option<ChainSpec<G>>, String>,
+		G: RuntimeGenesis {
```
Comment (bkchr) [inline core/cli/src/lib.rs]:
```suggestion
		G: RuntimeGenesis
	{
```

File: core/cli/src/lib.rs [hunk 10]
```diff
@@ -222,23 +213,294 @@ where
...
+
```
Comment (bkchr) [inline core/cli/src/lib.rs]:
```suggestion
```

File: core/cli/src/lib.rs [hunk 11]
```diff
@@ -222,23 +213,294 @@ where
...
+		self,
+		spec_factory: S
+	) -> error::Result<()>
+	where S: FnOnce(&str) -> Result<Option<ChainSpec<G>>, String>,
+		G: RuntimeGenesis {
+		
+		let config = create_config_with_db_path::<(), _, _>(spec_factory, &self.params.shared_params, self.version)?;
+		let db_path = config.database_path;
+
+		if self.params.yes == false {
```
Comment (bkchr) [inline core/cli/src/lib.rs]:
```suggestion
		if !self.params.yes {
```

Comment (tomaka) [reply core/cli/src/lib.rs]: That was already there, but yeah let's change that.

**Comment by tomaka** [comment]
> I'm not sure we want to have this code duplication in every run function.

I feel like the control flow is so much clearer that it is totally worth the code duplication.

> However you change the Service, it will result in a breaking change anyway.

For now I'm not breaking anything. Updating Polkadot can be tricky, so if possible I'm trying to not break stuff.

**Comment by bkchr** [comment]
I know that the current change does not break anything, but we will generate a new warning, that requires a new suppression or a switch to the new function.

I just want to make sure that we don't build any compatible intermediate state, just to not break interfaces. The current Service stuff is not well designed and I don't want that this bleeds into the new design in any way.

**Comment by bkchr** [comment]
(Feel free to merge this)

**Comment by tomaka** [comment]
> I know that the current change does not break anything, but we will generate a new warning, that requires a new suppression or a switch to the new function.

No, what I mean is that my work-in-progress service refactoring (which I haven't shown anyone yet) doesn't break anything.

What I'm doing is: I'm adding commits that introduce new stuff without breaking anything, then on top add commits that remove the old stuff. And then I'd open two PRs (one that adds this intermediary step that you don't like, and one that cleans up the stuff). This way we can upgrade Polkadot in a smoother way, and potentially revert the clean-up if it turns out that I forgot something.

**Comment by tomaka** [comment]
> (Feel free to merge this)

(I'm not in a rush, so will leave this PR a bit open in case more people want to comment on it)

**Comment by bkchr** [comment]
> > I know that the current change does not break anything, but we will generate a new warning, that requires a new suppression or a switch to the new function.
> 
> No, what I mean is that my work-in-progress service refactoring doesn't break anything.
> 
> What I'm doing is: I'm adding commits that introduce new stuff without breaking anything, then on top add commits that remove the old stuff. And then I'd open two PRs (one that adds this intermediary step that you don't like, and one that cleans up the stuff). This way we can upgrade Polkadot in a smoother way, and potentially revert the clean-up if it turns out that I forgot something.

If the intermediate state doesn't break anything, what is the reason to have it at all? The second pr will break stuff anyway and we will need to update Polkadot. 

I think having one concrete pr, will make it easier to iterate on the interface.

**Pull request merged**